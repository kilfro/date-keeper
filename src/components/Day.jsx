import {isSelected, isToday, isWeekend} from '../dateFunctions'

import React from 'react'
import {changeSelected} from '../store/actions'
import {connect} from 'react-redux'

const getStyle = (date, selected) => {
    let color = 'black'
    let backgroundColor = 'white'
    let border = '1px solid white'

    if (isWeekend(date)) {
        color = 'red'
    }

    if (isToday(date)) {
        border = `1px solid ${color}`
    }

    if (isSelected(date, selected)) {
        backgroundColor = color
        color = 'white'
    }

    return {
        height: '2rem',
        width: '2rem',
        margin: '0.4rem',
        fontSize: '1.3rem',
        fontWeight: 'bold',
        borderRadius: '50%',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        cursor: 'pointer',
        transitionDuration: '150ms',
        color,
        backgroundColor,
        border,
    }
}

export const Day = ({date, selected, changeSelected}) => {
    const changeHandler = () => changeSelected(date)

    return <div style={getStyle(date, selected)} onClick={changeHandler}>{date.getDate()}</div>
}

const mapStateToProps = state => ({
    selected: state.selected
})

const mapDispatchToProps = {
    changeSelected
}

export default connect(mapStateToProps, mapDispatchToProps)(Day)