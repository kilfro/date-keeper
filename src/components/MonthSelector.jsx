import React, {useState} from 'react'

import Hovered from './support/Hovered'
import Selector from './support/Selector'
import {connect} from 'react-redux'
import {setMonth} from '../store/actions'

const getMonthStyle = (isSelected) => ({
    color: isSelected ? 'red' : 'inherit',
    cursor: 'pointer',
    textAlign: 'center',
    padding: '0.7rem',
    transitionDuration: '200ms',
})

const hoveredMonthStyle = {
    backgroundColor: '#cbcbcb',
}

const months = [
    'январь', 'февраль', 'март',
    'апрель', 'май', 'июнь',
    'июль', 'август', 'сентябрь',
    'октябрь', 'ноябрь', 'декабрь',
]

const MonthSelector = ({date, setMonth}) => {
    const [showOptions, setShowOptions] = useState(false)

    const chooseAndClose = (month) => {
        setMonth(month)
        setShowOptions(false)
    }

    return (
        <Selector
            label={date.toLocaleString('ru', {month: 'long'})}
            labelColor={'red'}
            onChange={setShowOptions}
            showOptions={showOptions}
        >
            {months.map((month, index) =>
                <Hovered
                    style={getMonthStyle(index === date.getMonth())}
                    hoveredStyle={hoveredMonthStyle}
                    onClick={() => chooseAndClose(index)}
                    key={month}
                >
                    {month}
                </Hovered>)
            }
        </Selector>
    )
}

const mapDispatchToProps = {
    setMonth
}

export default connect(null, mapDispatchToProps)(MonthSelector)