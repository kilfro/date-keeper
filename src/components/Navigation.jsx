import { ChevronLeft, ChevronRight } from 'react-bootstrap-icons'
import { changeMonth, goToToday } from '../store/actions'

import Button from './support/Button'
import React from 'react'
import { connect } from 'react-redux'

const containerStyle = {
    display: 'grid',
    gridTemplateColumns: '1fr 1fr 1fr',
}

const Navigation = ({ goToToday, changeMonth }) => (
    <div style={containerStyle}>
        <Button onClick={() => changeMonth(-1)}><ChevronLeft /></Button>
        <Button onClick={goToToday}>сегодня</Button>
        <Button onClick={() => changeMonth(1)}><ChevronRight /></Button>
    </div>
)

const mapDispatchToProps = {
    goToToday,
    changeMonth,
}

export default connect(null, mapDispatchToProps)(Navigation)