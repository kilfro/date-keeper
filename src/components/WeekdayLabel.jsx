import React from 'react'

const getStyle = (label) => {
    let color = 'black'

    if (['СБ', 'ВС'].includes(label)) {
        color = 'red'
    }

    return {
        height: '2rem',
        width: '2rem',
        margin: '0.4rem',
        fontSize: '1.3rem',
        fontWeight: 'bold',
        border: '1px solid white',
        borderRadius: '50%',
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        color,
    }
}

const containerStyle = {
    display: 'flex',
    width: 'fit-content',
}

const WeekdayLabel = () => {
    const labels = ['ПН', 'ВТ', 'СР', 'ЧТ', 'ПТ', 'СБ', 'ВС']

    return (
        <div style={containerStyle}>
            {labels.map(label => <div style={getStyle(label)} key={label}>{label}</div>)}
        </div>
    )
}

export default WeekdayLabel