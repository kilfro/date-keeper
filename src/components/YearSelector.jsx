import { ChevronLeft, ChevronRight } from 'react-bootstrap-icons'
import React, { useState } from 'react'

import Button from './support/Button'
import Hovered from './support/Hovered'
import Selector from './support/Selector'
import { connect } from 'react-redux'
import { getYearSet } from '../dateFunctions'
import { setYear } from '../store/actions'

const getYearStyle = isSelected => ({
    padding: '1rem',
    cursor: 'pointer',
    color: isSelected ? 'red' : 'black',
})

const buttonStyle = {
    paddingTop: '0.7rem'
}

const YearSelector = ({ date, setYear }) => {
    const [showOptions, setShowOptions] = useState(false)
    const [basicYear, setBasicYear] = useState(date.getFullYear())

    const chooseAndClose = year => {
        setYear(year)
        setShowOptions(false)
        setBasicYear(year)
    }

    return (
        <Selector
            label={date.getFullYear()}
            onChange={setShowOptions}
            showOptions={showOptions}
        >
            <Button onClick={() => setBasicYear(basicYear - 9)} style={buttonStyle}><ChevronLeft /></Button>
            <div />
            <Button onClick={() => setBasicYear(basicYear + 9)} style={buttonStyle}><ChevronRight /></Button>

            {getYearSet(basicYear).map(year => (
                <Hovered
                    style={getYearStyle(year === date.getFullYear())}
                    hoveredStyle={{ backgroundColor: '#cbcbcb' }}
                    onClick={() => chooseAndClose(year)}
                    key={year}
                >
                    {year}
                </Hovered>
            ))}
        </Selector>
    )
}

const mapDispatchToProps = {
    setYear
}

export default connect(null, mapDispatchToProps)(YearSelector)