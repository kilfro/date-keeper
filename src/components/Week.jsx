import Day from './Day'
import PropTypes from 'prop-types'
import React from 'react'

const getStyle = (isFirst) => ({
    display: 'flex',
    width: 'fit-content',
    marginLeft: isFirst ? 'auto' : 0,
})

const Week = ({ days, isFirst }) => <div style={getStyle(isFirst)}>
    {days.map(day => <Day date={day} key={day.toString()} />)}
</div>

Week.propTypes = {
    days: PropTypes.arrayOf(PropTypes.object).isRequired,
    isFirst: PropTypes.bool,
}

export default Week