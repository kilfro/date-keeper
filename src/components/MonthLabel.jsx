import MonthSelector from './MonthSelector'
import React from 'react'
import YearSelector from './YearSelector'
import {connect} from 'react-redux'

const containerStyle = {
    margin: '0.5rem',
    fontSize: '2.5rem',
    textAlign: 'end',
    fontWeight: 'bold',
    display: 'flex',
    justifyContent: 'flex-end',
}

const MonthLabel = ({displayedMonth}) => {
    return (
        <div style={containerStyle}>
            <MonthSelector date={displayedMonth}/>
            <YearSelector date={displayedMonth}/>
        </div>
    )
}

const mapStateToProps = state => ({
    displayedMonth: state.displayedMonth
})

export default connect(mapStateToProps)(MonthLabel)