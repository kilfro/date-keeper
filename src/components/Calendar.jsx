import React, { useEffect } from 'react'
import { changeSelected, setDisplayedMonth, setOnChangeFunction } from '../store/actions'

import MonthLabel from './MonthLabel'
import Navigation from './Navigation'
import Week from './Week'
import WeekdayLabel from './WeekdayLabel'
import { connect } from 'react-redux'
import { getWeeks } from '../dateFunctions'

const style = {
    width: 'fit-content',
    fontFamily: 'sans-serif',
    userSelect: 'none',
    position: 'relative',
}

const Calendar = ({ displayedMonth, date, onChange, setOnChangeFunction, changeSelected, setDisplayedMonth }) => {
    useEffect(() => {
        setDisplayedMonth(date)
        changeSelected(date)
        setOnChangeFunction(onChange)
    }, [])

    return (
        <div style={style}>
            <MonthLabel />
            <Navigation />
            <WeekdayLabel />
            {getWeeks(displayedMonth).map((week, index) =>
                <Week
                    days={week}
                    key={week[0].toString()}
                    isFirst={index === 0}
                />)}
        </div>
    )
}

const mapStateToProps = state => ({
    displayedMonth: state.displayedMonth,
})

const mapDispatchToProps = {
    changeSelected,
    setOnChangeFunction,
    setDisplayedMonth,
}

export default connect(mapStateToProps, mapDispatchToProps)(Calendar)