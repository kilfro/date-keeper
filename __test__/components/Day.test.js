import { Day } from '../../src/components/Day'
import React from 'react'

describe('Day component', () => {
  test('should render correctly', () => {
    const props = {
      date: new Date(2020, 5, 1),
      selected: new Date(2021, 0, 1),
    }
    const component = shallow(<Day {...props} />)
    expect(component).toMatchSnapshot()
  })

  test("should render correctly if this day's selected", () => {
    const props = {
      date: new Date(2020, 5, 1),
      selected: new Date(2020, 5, 1),
    }
    const component = shallow(<Day {...props} />)
    expect(component).toMatchSnapshot()
  })

  test('should render correctly if this is today', () => {
    const props = {
      date: new Date(),
      selected: new Date(2020, 5, 1),
    }
    const component = shallow(<Day {...props} />)
    const border = component.get(0).props.style.border
    const [, , color] = border.split(' ')
    expect(color).toBe('black')
  })

  test('should render correctly if this is weekend', () => {
    const props = {
      date: new Date(2021, 2, 14),
      selected: new Date(),
    }
    const component = shallow(<Day {...props} />)
    const { color } = component.get(0).props.style
    expect(color).toBe('red')
  })
})
