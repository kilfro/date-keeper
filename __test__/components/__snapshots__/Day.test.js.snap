// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day component should render correctly 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "border": "1px solid white",
      "borderRadius": "50%",
      "color": "black",
      "cursor": "pointer",
      "display": "flex",
      "fontSize": "1.3rem",
      "fontWeight": "bold",
      "height": "2rem",
      "justifyContent": "center",
      "margin": "0.4rem",
      "transitionDuration": "150ms",
      "width": "2rem",
    }
  }
>
  1
</div>
`;

exports[`Day component should render correctly if this day's selected 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "black",
      "border": "1px solid white",
      "borderRadius": "50%",
      "color": "white",
      "cursor": "pointer",
      "display": "flex",
      "fontSize": "1.3rem",
      "fontWeight": "bold",
      "height": "2rem",
      "justifyContent": "center",
      "margin": "0.4rem",
      "transitionDuration": "150ms",
      "width": "2rem",
    }
  }
>
  1
</div>
`;
